Userland/Libraries/LibELF/DynamicObject.cpp

```diff
@@ -216,10 +216,39 @@ void DynamicObject::parse()
         m_size_of_relocation_table -= m_size_of_plt_relocation_entry_list;
     }
 
-    auto hash_section_address = hash_section().address().as_ptr();
-    // TODO: consider base address - it might not be zero
-    auto num_hash_chains = ((u32*)hash_section_address)[1];
-    m_symbol_count = num_hash_chains;
+    u32 const* hash_table_begin = (u32*)hash_section().address().as_ptr();
+
+    if (m_hash_type == HashType::SYSV) {
+        u32 n_chain = hash_table_begin[1];
+        m_symbol_count = n_chain;
+    } else {
+        // Determine amount of symbols by finding the chain with the highest
+        // starting index and walking this chain until the end to find the
+        // maximum index = amount of symbols.
+        using BloomWord = FlatPtr;
+
+        size_t const num_buckets = hash_table_begin[0];
+        size_t const num_omitted_symbols = hash_table_begin[1];
+        u32 const num_maskwords = hash_table_begin[2];
+        BloomWord const* bloom_words = (BloomWord const*)&hash_table_begin[4];
+        u32 const* const buckets = (u32 const*)&bloom_words[num_maskwords];
```

Comment:
Please try to use c++ style casts,
I also feel there has to be a more idiomatic way to do this

Reply:
fixed the casting :)